pkg/hive/cell/health.go

```diff
@@ -0,0 +1,240 @@
+	Stopped bool
+	// Final is the stopped message, if the module has been stopped.
+	Final string
+	// LastOK is the time of the last OK status update.
+	LastOK time.Time
+	// LastUpdated is the time of the last status update.
+	LastUpdated time.Time
+}
+
+// String returns a string representation of a Status, implements fmt.Stringer.
+func (s *Status) String() string {
+	var sinceLast string
+	if s.LastUpdated.IsZero() {
+		sinceLast = "never"
+	} else {
+		sinceLast = time.Since(s.LastUpdated).String() + " ago"
+	}
+	return fmt.Sprintf("Status{ModuleID: %s, Level: %s, Since: %s, Message: %s, Err: %v}",
+		s.ModuleID, s.Level, sinceLast, s.Message, s.Err)
+}
+
+// NewHealthProvider starts and returns a health status which processes
+// health status updates.
+func NewHealthProvider() Health {
+	p := &healthProvider{
+		moduleStatuses: make(map[string]Status),
+		running:        true,
+	}
+	return p
+}
+
+func (p *healthProvider) processed() uint64 {
+	return p.numProcessed.Load()
+}
+
+func (p *healthProvider) process(u Update) {
+	prev := func() Status {
+		p.mu.Lock()
+		defer p.mu.Unlock()
+
+		t := time.Now()
+		prev := p.moduleStatuses[u.ModuleID]
+
+		if !p.running {
+			return prev
+		}
+
+		ns := Status{
+			Update:      u,
+			LastUpdated: t,
+		}
+		switch u.Level {
+		case StatusOK:
+			ns.LastOK = t
+		case StatusStopped:
+			// If Stopped, set that module was stopped and preserve last known status.
+			ns = prev
+			ns.Stopped = true
+			ns.Final = u.Message
+		}
+		p.moduleStatuses[u.ModuleID] = ns
+		log.WithField("status", ns.String()).Debug("Processed new health status")
+		return prev
+	}()
+	p.numProcessed.Add(1)
+	if prev.Stopped {
+		log.Warnf("module %q reported health status after being Stopped", u.ModuleID)
+	}
+}
+
+// Finish stops the status provider, and waits for all updates to be processed or
+// returns an error if the context is cancelled first.
+func (p *healthProvider) Stop(ctx context.Context) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.running = false // following this, no new reporters will send.
+	return nil
+}
+
+// forModule returns a module scoped status reporter handle for emitting status updates.
+// This is used to automatically provide declared modules with a status reported.
+func (p *healthProvider) forModule(moduleID string) HealthReporter {
+	p.mu.Lock()
+	p.moduleStatuses[moduleID] = Status{Update: Update{
+		ModuleID: moduleID,
+		Level:    StatusUnknown,
+		Message:  "No status reported yet"},
+	}
+	p.mu.Unlock()
+
+	return &reporter{
+		moduleID: moduleID,
+		process:  p.process,
+	}
+}
+
+// All returns a copy of all the latest statuses.
+func (p *healthProvider) All() []Status {
+	p.mu.RLock()
+	all := maps.Values(p.moduleStatuses)
+	p.mu.RUnlock()
+	slices.SortFunc(all, func(a, b Status) bool {
+		return a.ModuleID < b.ModuleID
+	})
+	return all
+}
+
+// Get returns the latest status for a module, by module ID.
+func (p *healthProvider) Get(moduleID string) *Status {
```

Comment:
Could perhaps just return by value so this can't be misused to mutate the status?

Reply:
The map stores values, so this should return a pointer to a copied value.

(I think I want to avoid having to deal with a bunch of `_, ok := hr.Get(...)`)